parlai/agents/fairseq/fairseq.py

```diff
@@ -448,10 +451,31 @@ def eval_step(self, batch):
         if batch.label_vec is not None:
             # Interactive mode won't have a gold label
             self.trainer.valid_step(samples)
-        # Grade each of the candidate sequences
-        # TODO: grade everything in observations[i]['label_candidates']
 
-        if not self.args.skip_generation:
+        # Grade each of the candidate sequences
+        if batch.candidate_vecs is not None:
+            bsz = len(batch.text_vec)
+            best_cands = []
+            reranked_cands = []
+            # score the candidates for each item in the batch separately, so that
+            # we can support variable number of candidates
+            for i in range(bsz):
+                cands = batch.candidate_vecs[i]
+                ncand = len(cands)
+                # repeat the input many times
+                xs = batch.text_vec[i].unsqueeze(0).expand(ncand, -1)
+                # and appropriately pack the outputs
+                ys, _ = self._padded_tensor(cands)
+                s = self._make_sample(xs, ys)
+                # perform the actual grading
+                scored = list(self.scorer.score_batched_itr([s], cuda=self.use_cuda))
+                scores = [s[3][0]['score'].item() for s in scored]
+                ranked, = self._argsort(scores, batch.candidates[i], descending=True)
+                reranked_cands.append(ranked)
+                # best candidate is the highest scoring of the items
+                best_cands.append(ranked[0])
+            return Output(best_cands, reranked_cands)
```

Comment:
Should we support generation *and* ranking at the same time?

Reply:
I prefer that but it's probably not *that* useful